packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -288,6 +300,80 @@ class FuchsiaDevice extends Device {
     throwToolExit('No ports found running $isolateName');
     return null;
   }
+
+  FuchsiaIsolateDiscoveryProtocol  getIsolateDiscoveryProtocol(String isolateName) => FuchsiaIsolateDiscoveryProtocol(this, isolateName);
+}
+
+class FuchsiaIsolateDiscoveryProtocol {
+  FuchsiaIsolateDiscoveryProtocol(this._device, this._isolateName);
+
+  static const Duration _pollDuration = Duration(seconds: 10);
+  final Map<int, VMService> _ports = <int, VMService>{};
+  final FuchsiaDevice _device;
+  final String _isolateName;
+  final Completer<Uri> _foundUri = Completer<Uri>();
+  Timer _pollingTimer;
+  Status _status;
+  bool _locked = false;
+
+  Future<Uri> get uri {
+    _status ??= logger.startProgress(
+      'Waiting for a connection from $_isolateName on ${_device.name}...',
+      expectSlowOperation: true,
+    );
+    _pollingTimer ??= Timer.periodic(_pollDuration, _findIsolate);
+    return _foundUri.future;
+  }
+
+  void dispose() {
+    if (!_foundUri.isCompleted) {
+      _status?.cancel();
+      _status = null;
+      _foundUri.completeError(Exception('Did not complete'));
+    }
+  }
+
+  Future<void> _findIsolate(Timer timer) async {
+    // Prevent a timer callback from firing while there is already an operation
+    // pending
+    if (_locked) {
```

Comment:
Maybe use a regular timer with an explicit reschedule instead. Then you don't have to worry about this being re-entrant.

Reply:
Done